### PR TITLE
POL-723 Add Oracle Cloud Common Bill Ingestion

### PR DESCRIPTION
**Pull request by XOmniverse**
### Description

Adds a policy for automating the ingestion of Oracle Cloud cost & usage reports into the Flexera One platform via CBI.

### CODE REVIEW

File: cbi/oracle/oracle_cbi/oracle_cbi.pt [hunk 1]
```diff
@@ -0,0 +1,273 @@
+name "Oracle Cloud Common Bill Ingestion"
+rs_pt_ver 20180301
+type "policy"
+short_description "Downloads cost reports from Oracle Cloud (OCI) and then uploads them to a Flexera CBI endpoint. See the [README](https://github.com/flexera-public/policy_templates/tree/master/cbi/oracle/oracle_cbi) and [docs.flexera.com/flexera/EN/Automation](https://docs.flexera.com/flexera/EN/Automation/AutomationGS.htm) to learn more."
+severity "low"
+category "CBI"
```
Comment (nia-vf1) [inline cbi/oracle/oracle_cbi/oracle_cbi.pt:6]: Not sure where the options for category are defined, but agree this doesn't really fit under any of the existing categories. Could be `cost` maybe? 

Thoughts @bryankaraffa @taterdog590 ?

Comment (cdwilhelm) [reply cbi/oracle/oracle_cbi/oracle_cbi.pt:6]: @nia-vf1   my suggestion is to use Cost for the Category

**Comment by XOmniverse** [comment]
@cdwilhelm  @nia-vf1 (and to every other interested party)

Based on some testing and workarounds I had to do for Banco de Chile, I've fundamentally reworked this policy to run over multiple 15 minute iterations. This is to ensure that it doesn't hit memory or other limits within the Policy Engine.

The way the policy works now is the following:
* Check if an in-progress Bill Upload exists for the specified month (or current month if none is specified). If it does not, create one.
* See how many files have already been uploaded and compare it to the number of files sitting in Oracle for the month.
* If they are not equal, upload the next 20 (by default) files.
* If they are equal, upload no files and close the bill upload.

If the policy is run every 15 minutes, this should result in Oracle billing data being updated several times/day. The frequency can be increased to reduce how often this happens for customers with smaller amounts of data to ingest to avoid updating excessively frequently.

**Comment by nia-vf1** [comment]
> If the policy is run every 15 minutes, this should result in Oracle billing data being updated several times/day. The frequency can be increased to reduce how often this happens for customers with smaller amounts of data to ingest to avoid updating excessively frequently.

@XOmniverse - sorry for late reply, when you say frequency can be increased to reduce updates to ingest data, are you referring to the `param_commit_delay` parameter?

File: cost/oracle/oracle_cbi/oracle_cbi.pt [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+  description "The number of files in a single block."
+  default 20
+  min_value 1
+  max_value 20
+end
+
+parameter "param_commit_delay" do
+  type "number"
+  label "Commit Delay (Hours)"
+  description "Number of hours to wait between comitting bill uploads."
```
Comment (nia-vf1) [inline cost/oracle/oracle_cbi/oracle_cbi.pt:54]: Line 54 - minor change here - says 'comitting', should be 'committing'

File: cost/oracle/oracle_cbi/oracle_cbi.pt [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+rs_pt_ver 20180301
+type "policy"
+short_description "Downloads cost reports from Oracle Cloud (OCI) and then uploads them to a Flexera CBI endpoint. See the [README](https://github.com/flexera-public/policy_templates/tree/master/cost/oracle/oracle_cbi) and [docs.flexera.com/flexera/EN/Automation](https://docs.flexera.com/flexera/EN/Automation/AutomationGS.htm) to learn more."
+severity "low"
+category "Cost"
+default_frequency "15 minutes"
+info(
+  version: "2.0",
+  provider: "Oracle"
+)
```
Comment (nia-vf1) [inline cost/oracle/oracle_cbi/oracle_cbi.pt:11]: should have a `policy_set` and `service` in the info block

Not sure whether service applies here but maybe policy_set should be "Common Bill Ingest".
@cdwilhelm @bryankaraffa what are your thoughts?

File: cost/oracle/oracle_cbi/README.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+## Input Parameters
+
+This policy has the following input parameters required when launching the policy.
+
+- *Billing Period* - The year and month to process bills for in YYYY-MM format. Leave this parameter blank to always do the current month. Example: 2022-09
+- *Flexera CBI Endpoint* - The name of the Flexera CBI endpoint to use. Example: cbi-oi-oracle-oraclecloud
+- *Oracle Cloud Region* - The region of the Oracle Cloud Object Storage bucket containing the cost and usage reports. Example: us-phoenix-1
+- *Oracle Cloud Cost & Usage Bucket* - OCID of the Oracle Cloud Object Storage bucket containing the Cost and Usage reports.
+- *Block Size* - The number of files to upload with each execution of the policy. Changing from the default value of 20 is not recommended.
```
Comment (nia-vf1) [inline cost/oracle/oracle_cbi/README.md:22]: Think you are missing the Commit Delay parameter

**Review by bryankaraffa — COMMENTED** [review]
Combing backlog of reviews requested, is this is something you are trying to merge still ?

File: cost/oracle/oracle_cbi/oracle_cbi.pt (hunk 3 above)
Comment (XOmniverse) [reply cost/oracle/oracle_cbi/oracle_cbi.pt:11]: @cdwilhelm @bryankaraffa Thoughts on this? I can add them and leave them blank or put values in them.

**Comment by XOmniverse** [comment]
All but one of the pending issues should now be resolved.

I also added policy_set and service into the info block, setting both to "Common Bill Ingestion", but awaiting feedback on whether these are the correct values to use for a policy like this.

File: cost/oracle/oracle_cbi/oracle_cbi.pt (hunk 3 above)
Comment (cdwilhelm) [reply cost/oracle/oracle_cbi/oracle_cbi.pt:13]: policy_set Common Bill Ingest is good.  
I'm not sure service applies either.  you can leave blank or not include.  we can add it later if needed.

Comment (taniaSol) [reply cost/oracle/oracle_cbi/oracle_cbi.pt:13]: Agree on policy_set = "Common Bill Ingest"

**Review by nia-vf1 — APPROVED** [review]
LGTM

**Pull request merged**